src/enumerable.cr

```diff
@@ -965,6 +965,35 @@ module Enumerable(T)
     ary
   end
 
+  private def qselect_internal(data : Array(T), left : Int, right : Int, k : Int) : T
+    loop do
+      return data[left] if left == right
+      pivot_index = left + (right-left)//2
+      pivot_index = qselect_partition_internal(data, left, right, pivot_index)
+      if k == pivot_index
+        return data[k]
+      elsif k < pivot_index
+        right = pivot_index - 1
+      else
+        left = pivot_index + 1
+      end
+    end
+  end
+
+  private def qselect_partition_internal(data : Array(T), left : Int, right : Int, pivot_index : Int) : Int
+    pivot_value = data[pivot_index]
+    data.swap(pivot_index, right)
+    store_index = left
+    (left..right).each do |i|
```

Comment:
```suggestion
    (left..(right - 1)).each do |i|
```
After the swap, `pivot_value` is placed at `right`, so I suppose we can skip that index, correct?

Also, maybe `left.upto(right -1) do` reads a bit easier? Purely optional suggestion, just to get rid of the nested parenthesis. Semantically it's identical.

Reply:
Actually, an exclusive range would be best👌 (assuming it's semantically correct)
```suggestion
    (left...right).each do |i|
```